src/sema/types.rs

```diff
@@ -1503,25 +1607,28 @@ impl Type {
 
     /// Does this type contain any types which are variable-length
     pub fn is_dynamic(&self, ns: &Namespace) -> bool {
-        match self {
+        self.is_dynamic_internal(ns, &mut HashSet::new())
+    }
+
+    fn is_dynamic_internal(&self, ns: &Namespace, structs_visited: &mut HashSet<usize>) -> bool {
+        self.guarded_recursion(structs_visited, false, |structs_visited| match self {
```

Comment:
Why do you need the recursion guard if you already identify the recursive fields beforehand?

Reply:
A recursive field does not necessarily imply  a dynamic field.

Reply:
I mean, can you avoid the recursion guard if you check the `recursive` boolean?

Reply:
I don't think so.  Given the following:

```solidity
struct A {  A[] a; }
```

Now, `is_dynamic` should return true for struct `A`. But if we avoid recursion because field `A.a` is recursive, won't find out.

Reply:
In the example you've given, you can detect the dynamic array in `Type::Array` before recursing for A. In this case, the recursion can be avoided.

Reply:
It's not possible.

```solidity
struct A {  B b; }
struct B { C[] c; }
struct C { C c; }
```

 `is_dynamic` should stilll return true for struct A. We need to do the recursion, regardless whether a struct field is flagged as recursive